crates/ark/src/variables/variable.rs

```diff
@@ -918,132 +931,216 @@ impl PositronVariable {
-                    //       maybe use anyhow here instead ?
-                    let row_index = path_element.parse::<isize>().unwrap();
+        // R6 objects may be accessed with special elements called <methods> and <private>.
+        // For them, we'll have to build the next node artifically.
+        if r_inherits(object.sexp, "R6") && access_key.starts_with("<") {
+            return Ok(EnvironmentVariableNode::R6Node {
+                object,
+                name: access_key.clone(),
+            });
+        }
 
-                    EnvironmentVariableNode::VectorElement {
+        match r_typeof(*object) {
+            ENVSXP => Self::get_envsxp_child_node_at(object, access_key),
+            VECSXP | EXPRSXP => {
+                let index = parse_index(access_key)?;
+                Ok(EnvironmentVariableNode::Concrete {
+                    object: RObject::view(harp::list_get(object.sexp, index)),
+                })
+            },
+            LISTSXP => {
+                let mut pairlist = *object;
+                let index = parse_index(access_key)?;
+                for _i in 0..index {
+                    pairlist = unsafe { CDR(pairlist) };
+                }
+                Ok(EnvironmentVariableNode::Concrete {
+                    object: RObject::view(unsafe { CAR(pairlist) }),
+                })
+            },
+            LGLSXP | RAWSXP | STRSXP | INTSXP | REALSXP | CPLXSXP => {
+                if r_is_matrix(object.sexp) {
+                    Ok(EnvironmentVariableNode::Matrixcolumn {
                         object,
-                        index: n_row * index + row_index,
-                    }
-                },
-            }
+                        index: parse_index(access_key)?,
+                    })
+                } else {
+                    Ok(EnvironmentVariableNode::AtomicVectorElement {
+                        object,
+                        index: parse_index(access_key)?,
+                    })
+                }
+            },
+            _ => Err(harp::Error::Anyhow(anyhow!(
+                "Unexpected child at {access_key}"
+            ))),
+        }
+    }
+
+    fn get_child_node_at(
+        node: EnvironmentVariableNode,
+        path_elt: &String,
+    ) -> harp::Result<EnvironmentVariableNode> {
+        match node {
+            EnvironmentVariableNode::Concrete { object } => {
+                Self::get_concrete_child_node(object, path_elt)
+            },
+
+            EnvironmentVariableNode::R6Node { object, name } => {
+                match name.as_str() {
+                    "<private>" => {
+                        let env = Environment::new(object);
+                        let enclos = Environment::new(RObject::view(env.find(".__enclos_env__")?));
+                        let private = Environment::new(RObject::view(enclos.find("private")?));
+
+                        // TODO: it seems unlikely that private would host active bindings
```

Comment:
I don't remember, does `find()` return the active binding function? I vaguely remember it depends on the R version, with older versions returning the function and newer returning the value.

It seems like this could all be cleaned up by using safer / more explicit accessors, because this code currently raises lots of questions.

Reply:
I'm not sure too, this piece of code is not new thugh. Should we change it a follow up PR, maybe just rewriting in R: https://github.com/posit-dev/ark/pull/561#discussion_r1810922083